### ISSUE TITLE
Inconsistent behavior with and without distributed scope

### ISSUE THREAD

**Issue by anirudhraju** [tensorflow/tensorflow#30961]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Darwin Kernel Version 16.7.0
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): installed from pypi
- TensorFlow version (use command below):  v2.0.0-beta0-16-g1d91213fe7 2.0.0-beta1
- Python version: 3.7.3
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: -

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Training on mac os, with CPU only, with and without distributed scope for mnist example. The loss and accuracy is incorrect with distributed scope turned off

With distributed scope on:
Train on 550.0 steps, validate on 50 steps
Epoch 1/5
550/550 [==============================] - 6s 10ms/step - loss: 4.5770 - accuracy: 0.7565 - val_loss: 1.0736 - val_accuracy: 0.7990
Epoch 2/5
550/550 [==============================] - 2s 3ms/step - loss: 0.7952 - accuracy: 0.8072 - val_loss: 0.6584 - val_accuracy: 0.8282
Epoch 3/5
550/550 [==============================] - 2s 3ms/step - loss: 0.5272 - accuracy: 0.8367 - val_loss: 0.5347 - val_accuracy: 0.8438
Epoch 4/5
550/550 [==============================] - 2s 3ms/step - loss: 0.4477 - accuracy: 0.8501 - val_loss: 0.4830 - val_accuracy: 0.8572
Epoch 5/5
550/550 [==============================] - 2s 3ms/step - loss: 0.4037 - accuracy: 0.8591 - val_loss: 0.4780 - val_accuracy: 0.8456

Without distributed scope:
Epoch 1/5
550/550 [==============================] - 3s 6ms/step - loss: 14.5062 - accuracy: 0.1000 - val_loss: 14.4869 - val_accuracy: 0.1012
Epoch 2/5
550/550 [==============================] - 2s 4ms/step - loss: 14.5080 - accuracy: 0.0999 - val_loss: 14.4869 - val_accuracy: 0.1012
Epoch 3/5
550/550 [==============================] - 2s 4ms/step - loss: 14.5080 - accuracy: 0.0999 - val_loss: 14.4869 - val_accuracy: 0.1012
Epoch 4/5
550/550 [==============================] - 2s 4ms/step - loss: 14.5080 - accuracy: 0.0999 - val_loss: 14.4869 - val_accuracy: 0.1012
Epoch 5/5
550/550 [==============================] - 2s 4ms/step - loss: 14.5080 - accuracy: 0.0999 - val_loss: 14.4869 - val_accuracy: 0.1012

**Describe the expected behavior**
The training is expected to work correctly even with distributed scope turned off. This was working correctly in the tf2.0 alpha release but is an issue in the tf2.0 beta1 release. 

**Code to reproduce the issue**
```python
import tensorflow as tf
import contextlib
import sys

distributed=False
if len(sys.argv) > 1 and sys.argv[1] == "distributed":
    distributed=True
    print("asdf enabled distributed trainer")

class NoOpScope:
    def scope(self):
        return contextlib.suppress()

distribution_strategy = tf.distribute.MirroredStrategy() if distributed else NoOpScope()

with distribution_strategy.scope():
    model = tf.keras.models.Sequential()
    model.add(tf.keras.layers.Flatten(input_shape=[28, 28]))
    model.add(tf.keras.layers.Dense(300, activation="relu"))
    model.add(tf.keras.layers.Dense(100, activation="relu"))
    model.add(tf.keras.layers.Dense(10, activation="softmax"))
    model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(), optimizer=tf.keras.optimizers.Adam(), metrics=['accuracy'])

(x_train_full, y_train_full), (x_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()
x_dev, x_train = x_train_full[:5000], x_train_full[5000:]
y_dev, y_train = y_train_full[:5000], y_train_full[5000:]
train_data = tf.data.Dataset.from_tensor_slices((x_train, y_train)).shuffle(55000).repeat().batch(100)
dev_data = tf.data.Dataset.from_tensor_slices((x_dev, y_dev)).batch(100)

model.fit(train_data, 
          epochs=5,
          steps_per_epoch=55000/100,
          validation_data=dev_data)
```

Please copy to a script and run as:
Without distributed scope: python3 script.py
With distributed scope : python3 script.py distributed

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gadagashwini** [comment]
@anirudhraju Looks like the code is incomplete. Please provide the complete code to reproduce the issue. Thanks!

**Comment by anirudhraju** [comment]
Reformatted the above code blurb with python markdown for readability. Please copy to a script and run as:
Without distributed scope: python3 script.py
With distributed scope : python3 script.py distributed

**Comment by gadagashwini** [comment]
I am able to reproduce the issue on Colab with Tensorflow version 2.0.0.beta1. Please find the [gist](https://colab.research.google.com/drive/1NsqVwxKeBy6zZOA4KCC0etgzy7NUZYy9) of colab. Thanks!

**Comment by guptapriya** [comment]
I believe you are missing this one step in the input pipeline:
`x_train_full = x_train_full / 255.0` 
After adding this, the code without strategy seems to work fine too. 

See this tutorial for a very similar example: 
https://www.tensorflow.org/beta/tutorials/keras/basic_classification

I don't understand though why
- with strategy it seems to work without needing this extra step ( i will look into that) 
- why it was working for you before beta. (are you sure you didn't have that division at the time?)

**Comment by anirudhraju** [comment]
>why it was working for you before beta. (are you sure you didn't have that division at the time?)
Yes, this works ok in TF2.0 alpha0 version without the distributed scope, and without the division by 255 operation. 
In beta, the behavior is inconsistent with and without the distributed scope

**Comment by anirudhraju** [comment]
@qlzh727  Any update on this issue? Do we know why this behavior is consistent between tf2.0 alpha and beta?

**Comment by qlzh727** [comment]
Thanks for the notice. let me take a look today.

**Comment by qlzh727** [comment]
I wasn't able to reproduce the issue with the snippet above on the latest head. The issue seems to be already fixed by the new execution path.

**Comment by qlzh727** [comment]
Closing this issue due to the lack of activity. Feel free to reopen it if you still see the issue.
